builtin/v13/migration/miner.go

```diff
@@ -0,0 +1,199 @@
+package migration
+
+import (
+	"context"
+	"sync"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-amt-ipld/v4"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin/v13/miner"
+	"github.com/filecoin-project/go-state-types/builtin/v13/util/adt"
+
+	miner12 "github.com/filecoin-project/go-state-types/builtin/v12/miner"
+	"github.com/filecoin-project/go-state-types/migration"
+)
+
+type providerSectors struct {
+	lk sync.Mutex
+
+	// Populated ONLY on the first run-through, when there's no cache to diff against
+	dealToSector         map[abi.DealID]abi.SectorID
+	minerToSectorToDeals map[abi.ActorID]map[abi.SectorNumber][]abi.DealID
+
+	// Populated ONLY on future run-throughs if there is a cache to diff against
+	newDealsToSector              map[abi.DealID]abi.SectorID
+	updatesToMinerToSectorToDeals map[abi.ActorID]map[abi.SectorNumber][]abi.DealID
+}
+
+// minerMigration is technically a no-op, but it collects a cache for market migration
+type minerMigrator struct {
+	providerSectors *providerSectors
+
+	OutCodeCID cid.Cid
+}
+
+func newMinerMigrator(ctx context.Context, store cbor.IpldStore, outCode cid.Cid, ps *providerSectors) (*minerMigrator, error) {
+	return &minerMigrator{
+		providerSectors: ps,
+
+		OutCodeCID: outCode,
+	}, nil
+}
+
+func (m *minerMigrator) MigrateState(ctx context.Context, store cbor.IpldStore, in migration.ActorMigrationInput) (result *migration.ActorMigrationResult, err error) {
+	var inState miner12.State
+	if err := store.Get(ctx, in.Head, &inState); err != nil {
+		return nil, xerrors.Errorf("failed to load miner state for %s: %w", in.Address, err)
+	}
+
+	ctxStore := adt.WrapStore(ctx, store)
+
+	mid, err := address.IDFromAddress(in.Address)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get miner ID: %w", err)
+	}
+
+	inSectors, err := adt.AsArray(ctxStore, inState.Sectors, miner12.SectorsAmtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load sectors array: %w", err)
+	}
+
+	hasCached, prevSectors, err := in.Cache.Read(migration.MinerPrevSectorsInKey(in.Address))
```

Comment:
We need to think about use of the cache -- providerSectors is populated based on whether or not there's a cache hit, but the market migration doesn't consider the possibility that some miners were cached, but others weren't.